torch/_dynamo/variables/builtin.py

```diff
@@ -1214,10 +1214,41 @@ def _grad_changed(old, new):
                                 grapharg.example.grad = torch.zeros(
                                     grad_shape_specialized, device=new_grad.device
                                 )
-                            else:
-                                grapharg.example.grad = None
-                        return VariableBuilder(tx, source)(grapharg.example.grad)
 
+                        # Check if grad changed via direct setting, such as an inplace op, see
+                        # https://github.com/pytorch/pytorch/pull/112811
+                        if _grad_changed(old_grad, new_grad):
```

Comment:
I don't understand this implementation strategy. (This is also my fault for punting on a close review on https://github.com/pytorch/pytorch/pull/112811)

The previous PR observes that we have a problem whereby the real tensor grad attribute (grapharg.example) is desynchronized with the fake tensor (meta["example_value"].grad), but the old code directly grabbed grad off of the real tensor grad (grapharg.example). It says "We have no proper mechanism for updating the hint / grapharg.example (the real value in (2) above) midway through trace", but in fact, we MUST NOT update grapharg.example midway through the trace, as that would be modifying the real CPython env, which we are not allowed to during symbolic tracing.

To me, the logical next conclusion when this happens is "Well, that means that we should just pull the grad off of the fake tensor instead!" The previous PR seems to observe this is a possibility, but says ".grad access handling in dynamo peeks at the underlying value, the real tensor, because re-piping FakeTensors already made with this fake_mode through builder anew is a no go." Well, there is certainly some care you have to take when doing this, but it seems to me that we want to do *something* with the already existing TensorVariable.

I think the optimal strategy looks something like this:

1. We want to lazily create the TensorVariable for grad, because frequently we will be compiling code that doesn't care about what grad is on tensor, so we don't want to muck around with it unnecessarily
2. The TensorVariable for grad should be stored on the base TensorVariable, the logical place to do it
3. We probably shouldn't immediately fakeify the grad when we are doing fake tensorification. In fact, we should never consult the grad field on the fake tensor (I think this is fine, so long as we don't have any operations that directly try to access grad on fake tensors. I think grad accumulate is the only suspicious one.)
4. If a mutation happens to TensorVariable's grad before we've populated its grad, we should go ahead and remember it on the base. We ONLY build the grad if the lazy tensor variable we've put in its spot has survived until this point in time.

WDYT?